### PR TITLE
firefox: enable RLBox sandboxing

### PR DESCRIPTION
**Pull request by mweinelt**
https://hacks.mozilla.org/2021/12/webassembly-and-back-again-fine-grained-sandboxing-in-firefox-95/

###### Description of changes

Firefox since 95.0 supports sandboxing based on a  webassembly project called RLBox. Building it requires sysroot support for wasi32 cross builds and we haven't had any luck in getting that to work.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Fails to link wasm32 binaries.

```
CMake Error at /nix/store/sgsvy949kd2sr2vsxcffz5jkv855rmib-cmake-3.22.2/share/cmake-3.22/Modules/CMakeTestCCompiler.cmake:69 (message):
  The C compiler

    "/nix/store/lmnvxwbzh8nbxls0i8604bjk6pv5fygg-wasm32-unknown-wasi-clang-wrapper-8.0.1/bin/wasm32-unknown-wasi-clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/nix/store/is8wg3j77z9ndpsvkm85sfbrlvvqd66l-gnumake-4.3/bin/make -f Makefile cmTC_387d9/fast && /nix/store/is8wg3j77z9ndpsvkm85sfbrlvvqd66l-gnumake-4.3/bin/make  -f CMakeFiles/cmTC_387d9.dir/build.make CMakeFiles/cmTC_387d9.dir/build
    make[1]: Entering directory '/build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp'
    Building C object CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj
    /nix/store/lmnvxwbzh8nbxls0i8604bjk6pv5fygg-wasm32-unknown-wasi-clang-wrapper-8.0.1/bin/wasm32-unknown-wasi-clang    -MD -MT CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj -MF CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj.d -o CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj -c /build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp/testCCompiler.c
    clang-8: warning: argument unused during compilation: '-rtlib=compiler-rt' [-Wunused-command-line-argument]
    Linking C executable cmTC_387d9
    /nix/store/sgsvy949kd2sr2vsxcffz5jkv855rmib-cmake-3.22.2/bin/cmake -E cmake_link_script CMakeFiles/cmTC_387d9.dir/link.txt --verbose=1
    /nix/store/lmnvxwbzh8nbxls0i8604bjk6pv5fygg-wasm32-unknown-wasi-clang-wrapper-8.0.1/bin/wasm32-unknown-wasi-clang CMakeFiles/cmTC_387d9.dir/testCCompiler.c.obj -o cmTC_387d9 
    wasm32-unknown-wasi-wasm-ld: error: /nix/store/fi3903r6zj9psp8c9ry6llq0lw0j9fv2-wasilibc-static-wasm32-unknown-wasi-unstable-2021-09-23/lib/libc.a(dlmalloc.o): Bad section type
    clang-8: error: lld command failed with exit code 1 (use -v to see invocation)
    make[1]: *** [CMakeFiles/cmTC_387d9.dir/build.make:100: cmTC_387d9] Error 1
    make[1]: Leaving directory '/build/libcxxabi-8.0.1.src/build/CMakeFiles/CMakeTmp'
    make: *** [Makefile:127: cmTC_387d9/fast] Error 2
```

**Comment by mweinelt** [comment]
The simple test program works with clang 11, but then it fails to link later.

```
wasm-ld: error: unable to find library -lc++
wasm-ld: error: unable to find library -lc++abi
```

**Comment by mweinelt** [comment]
I've tried `SYSROOT`, `LIBRARY_PATH` `WASM_CXXFLAGS` and now `NIX_CLAGS_LINK`. I cannot get the linker to find those two libraries, which are located below `${wasiSysRoot}/lib`.

```
[...]
security/rlbox/rlbox.wasm
rm -f rlbox.wasm
/nix/store/phf9n58asd4yas9dlmaxvnlbqdbksk48-wasm-clang++ -std=gnu++17 --sysroot=/nix/store/hjbqjl31hv4999acqm1jvvaka089q9kp-wasi-sysroot -o rlbox.wasm -Wl,--export-all -Wl,--stack-first -Wl,-z,stack-size=262144 -Wl,--no-entry -Wl,--growable-table ogg_alloc.wasm ogg_bitwise.wasm ogg_framing.wasm xmlparse.wasm xmlrole.wasm xmltok.wasm wasm2c_sandbox_wrapper.wasm mozHunspellRLBoxSandbox.wasm affentry.wasm affixmgr.wasm csutil.wasm hashmgr.wasm hunspell.wasm phonet.wasm replist.wasm suggestmgr.wasm GraphiteExtra.wasm CmapCache.wasm Code.wasm Collider.wasm Decompressor.wasm Face.wasm FeatureMap.wasm FileFace.wasm Font.wasm GlyphCache.wasm GlyphFace.wasm Intervals.wasm Justifier.wasm NameTable.wasm Pass.wasm Position.wasm Segment.wasm Silf.wasm Slot.wasm Sparse.wasm TtfUtil.wasm UtfCodec.wasm call_machine.wasm gr_char_info.wasm gr_face.wasm gr_features.wasm gr_font.wasm gr_logging.wasm gr_segment.wasm gr_slot.wasm json.wasm RLBoxWOFF2Sandbox.wasm table_tags.wasm variable_length.wasm woff2_common.wasm woff2_dec.wasm woff2_out.wasm -lwasi-emulated-process-clocks
wasm-ld: error: unable to find library -lc++
wasm-ld: error: unable to find library -lc++abi
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [/build/firefox-97.0.2/config/rules.mk:498: rlbox.wasm] Error 1
make[3]: Leaving directory '/build/firefox-97.0.2/obj-x86_64-pc-linux-gnu/security/rlbox'
make[2]: *** [/build/firefox-97.0.2/config/recurse.mk:72: security/rlbox/target-objects] Error 2
make[2]: *** Waiting for unfinished jobs....
```

File: pkgs/development/libraries/wasilibc/default.nix [hunk 1]
```diff
@@ -1,31 +1,46 @@
...
   };
+
+  # clang-13: error: argument unused during compilation: '-rtlib=compiler-rt' [-Werror,-Wunused-command-line-argument]
+  postPatch = ''
+    substituteInPlace Makefile \
+      --replace "-Werror" ""
+  '';
+
+  NIX_CFLAGS_COMPILE = [
+    "-I/build/source/sysroot/include"
```
Comment (Mic92) [inline pkgs/development/libraries/wasilibc/default.nix:22]: That will break if someone has no build sandbox enabled.
Maybe do something like:

```
preBuild = ''
  NIX_CFLAGS_COMPILE=$(pwd)/sysroot/include
'';
```

**Comment by mweinelt** [comment]
Builds for me on x86_64-linux and hunspell is still working, which should be an indicator that we didn't break things that could be inside the sandbox.

**Comment by mweinelt** [comment]
@matthewbauer I'm not sure why you weren't requested for review, but I touched your wasilibc package in this change and upgrade the llvm version used for wasm32. Hope you're fine with the changes.

**Pull request merged**